### PR TITLE
Final Exam - Dijana Damchevska 5048

### PR DESCRIPTION
**Pull request by damcevskad**
Dijana Damchevska 5048

### CODE REVIEW

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-create/band-create.component.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  bandForm: FormGroup;
+  genres: string[] = [];
+
+  places$: Observable<Place[]> = this.bandService.getPlaces();
+
+  constructor(
+    private formBuilder: FormBuilder,
+    private bandService: BandService,
+    private router: Router
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-create/band-create.component.ts:23]: +5 points for injecting and using `Router`

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-create/band-create.component.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    this.bandForm = this.formBuilder.group({
+      name: ['', Validators.required],
+      genre: ['', [Validators.required]],
+      formed: ['', [Validators.required]],
+      location: ['', Validators.required],
+    });
+  }
+
+  ngOnInit(): void {
+    this.bandService.getGenre().subscribe(genres => {
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-create/band-create.component.ts:34]: +5 points for using service observables

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<div class="band-details-container" *ngIf="!!band">
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.html:1]: +5 points for the details component

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      if (a.year !== b.year) {
+        return a.year - b.year;
+      }
+      return a.name.localeCompare(b.name);
+    });
+  }
+  return [];
+}
+
+calculateYearsActive(): void {
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts:80]: Extra +5 for the years active

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  templateUrl: './band-details.component.html',
+  styleUrl: './band-details.component.css'
+})
+export class BandDetailsComponent {
+
+  band?: Band;
+  yearsActive: string = "";
+
+  constructor(
+    private route: ActivatedRoute,
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts:18]: +5 points for injecting and using `ActivatedRoute`

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  yearsActive: string = "";
+
+  constructor(
+    private route: ActivatedRoute, 
+    private bandService: BandService,
+    private router: Router,
+  ) { }
+
+  ngOnInit(): void {
+    this.route.paramMap.pipe(
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts:24]: +5 points for using routing observables

Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts:24]: Extra +5 points for using `pipe`

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      map(params => params.get('id')),
+      filter(id => this.isNumeric(id)),
+      map(id => parseInt(id!, 10)),
+      switchMap(id => this.bandService.getBand(id)),
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-details/band-details.component.ts:28]: Extra +5 points for using `switchMap`

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-edit/band-edit.component.html [hunk 8]
```diff
@@ -0,0 +1,33 @@
+<div class="edit-form-container" *ngIf="!!band">
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-edit/band-edit.component.html:1]: There is a mismatch between the formControls in the markup and the component code, so this form just does not work - it should be relatively simple fix.
+3 points for an almost functional edit component

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-edit/band-edit.component.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  styleUrl: './band-edit.component.css'
+})
+export class BandEditComponent {
+
+
+  band?: Band;
+  editForm: FormGroup = new FormGroup({});
+
+  genres$: Observable<string[]> = this.bandsService.getGenre();
+  places$: Observable<Place[]> = this.bandsService.getPlaces();
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-edit/band-edit.component.ts:20]: Extra +5 for using the `async` pipe

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.html [hunk 10]
```diff
@@ -0,0 +1,51 @@
+<div class="band-list">
+  <div class="add">
+    <button (click)="addBand()">Add Band</button>
+  </div>
+
+  <table>
+    <thead>
+      <tr>
+        <th (click)="updateSorting('id')">ID</th>
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.html:9]: Extra +5 for sorting

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.html [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        <td>
+          <button class="action-button" (click)="viewBand(band)">View</button>
+          <button class="action-button" (click)="editBand(band)">Edit</button>
+          <button class="action-button" (click)="deleteBand(band)">Delete</button>
+        </td>
+      </tr>
+    </tbody>
+  </table>
+
+  <dialog #dialog id="modal">
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.html:37]: Extra +5 for dialog implementation

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.html [hunk 12]
```diff
@@ -0,0 +1,51 @@
+<div class="band-list">
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.html:1]: +5 points for list markup

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import { Router } from '@angular/router';
+import { Subscription } from 'rxjs';
+import { Band } from '../../models/band';
+
+@Component({
+  selector: 'app-band-list',
+  templateUrl: './band-list.component.html',
+  styleUrl: './band-list.component.css'
+})
+export class BandListComponent  implements OnInit, OnDestroy{
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.ts:12]: +5 points for list component code

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+  ngOnInit(): void {
+    this.bandsSubscription$ = this.bandsService.getBands().subscribe({
+      next: bands => this.bands = bands,
+      error: err => console.log(err)
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.bandsSubscription$?.unsubscribe();
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/band-list/band-list.component.ts:35]: +5 point for using `unsubscribe`

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/header/header.component.html [hunk 15]
```diff
@@ -0,0 +1,9 @@
+<nav class="header">
+  <h1>Metal Head</h1>
+  <ul class="nav-list">
+    <li><button [routerLink]="['/bands']">Bands</button></li>
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/header/header.component.html:4]: +5 point for functional header component

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/footer/footer.component.html [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<footer class="footer">
+    <p>
+      <br />
+      {{ student?.id }} - {{ student?.name }}
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/footer/footer.component.html:4]: +5 point for functional footer component

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/statistics/statistics.component.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+import { BandService } from '../../services/band/band.service';
+
+@Component({
+  selector: 'app-statistics',
+  templateUrl: './statistics.component.html',
+  styleUrl: './statistics.component.css'
+})
+
+export class StatisticsComponent implements OnInit {
+  totalBands: number = 0;
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/components/statistics/statistics.component.ts:11]: +5 point for functional statistics component

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/models/band.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
+export interface Band {
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/models/band.ts:1]: +5 points for the models, the definitions are ok.

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/models/mapping.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
+import { Band, Album, Place } from "./band";
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/models/mapping.ts:1]: The point of the mapping is to transform the data to a better format. In this case both formats are the same, so the transformation is in place. (I.e. the frontend could have split the location into city and country in the mapping) Still, the code is ok, +5 points

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+      tap(band => {
+        const bands = this.bandSubject$.getValue();
+        bands.push(band);
+        this.bandSubject$.next(bands);
+      }),
+    );
+  }
+
+getPlaces(): Observable<Place[]> {
+  return this.getBands().pipe(
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts:73]: There was a `places` thing in the api, that should have been used 😄

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+})
+export class BandService {
+
+  private bandSubject$ = new BehaviorSubject<Band[]>([]);
+
+  constructor(private http: HttpClient) { }
+
+  getBands() {
+    return this.http.get<BandResponse[]>(`${BASE_URL}/bands`).pipe(
+      map(bandResponses => bandResponses.map(bandResponse => toBand(bandResponse))),
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts:21]: +5 points for state management with a subject

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+const BASE_URL = "http://localhost:3000"
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BandService {
+
+  private bandSubject$ = new BehaviorSubject<Band[]>([]);
+
+  constructor(private http: HttpClient) { }
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts:17]: +5 points for correct `HttpClient` usage

File: finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+import { Band, Place } from '../../models/band';
+import { BandResponse } from '../../models/server';
+import { toBand, toBandResponse } from '../../models/mapping';
+
+const BASE_URL = "http://localhost:3000"
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BandService {
```
Comment (sweko) [inline finals-2024-01-16/code/5048-dijana-damcevska/src/app/services/band/band.service.ts:13]: +10 points for the service design

**Review by sweko — APPROVED** [review]
@damcevskad - 151/100 points - Simply amazing. It has been a privilege

**Comment by damcevskad** [comment]
Thank you professor. It has been a pleasure learning from you.

**Pull request merged**